### PR TITLE
SC1/SC2 Recipe

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/recipes/starcoder.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+    # Optimizer
+    warmup_steps=500,
+    constant_steps=0,
+    min_lr=3e-5,
+    max_lr=3e-4,
+    # Training function
+    fn=pretrain,
+) -> run.Partial:
+    """
+    Create a pre-training recipe for Starcoder2 3b model.
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/starcoder.py:175]: typo in model name

File: nemo/collections/llm/recipes/starcoder.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    precision: str = "bf16-mixed",
+    accumulate_grad_batches: int = 1,
+    limit_test_batches: int = 32,
+    limit_val_batches: int = 32,
+    log_every_n_steps: int = 10,
+    val_check_interval: int = 2000,
+    callbacks: Optional[list[run.Config[Callback]]] = None,
+) -> run.Config[nl.Trainer]:
+    """
+    Configure the NeMo Lightning Trainer for Starcoder2 models.
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/starcoder.py:75]:
```suggestion
    Configure the NeMo Lightning Trainer for Starcoder models.
```

File: nemo/collections/llm/recipes/starcoder.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+@run.cli.factory(target=finetune, name=NAME)
+def finetune_recipe(
+    dir: Optional[str] = None,
+    name: str = "default",
+    num_nodes: int = 1,
+    num_gpus_per_node: int = 8,
+    peft_scheme: Optional[str] = 'lora',
+) -> run.Partial:
+    """
+    Create a fine-tuning recipe for Starcoder2 3B model.
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/starcoder.py:272]: references to starcoder 2 3b in this doc string

File: nemo/collections/llm/recipes/starcoder2_15b.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    Note:
+        This recipe uses the SQuAD dataset for fine-tuning. For more information
+        on fine-tuning LLMs with NeMo, see the fine-tuning guide in the
+        `examples/llm/finetune/` directory.
+    """
+    recipe = default_finetune_recipe(model(), "bigcode/starcoder2-15b", dir, name, num_nodes, num_gpus_per_node)
+    if peft_scheme is None or peft_scheme.lower() == 'none':
+        recipe.trainer.strategy.tensor_model_parallel_size = 4
+        recipe.optim.config.lr = 5e-6
+    elif peft_scheme.lower() == 'lora':
```
Comment (cuichenx) [inline nemo/collections/llm/recipes/starcoder2_15b.py:218]: Does 15B lora work with TP1 PP1? I thought you would need TP2. Same question for full finetuning TP4

Comment (suiyoubi) [reply nemo/collections/llm/recipes/starcoder2_15b.py:218]: tested TP1PP1 works with 15B loRA, and TP4works for SFT

**Review by cuichenx — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]
LGTM

**Pull request merged**